File: interval/insert-interval/InsertInterval.java
// https://leetcode.com/problems/insert-interval/?envType=study-plan-v2&envId=top-interview-150

// Time complexity: O(n) where n is the number of intervals
// Space complexity: O(n) where n is the number of intervals
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {

        ArrayList<int[]> outputIntervals = new ArrayList<>();
        boolean inserted = false;
        int newIntervalStart = newInterval[0];
        int newIntervalEnd = newInterval[1];

        for(int[] interval: intervals) {

            int currentIntervalStart = interval[0];
            int currentIntervalEnd = interval[1];

            if(currentIntervalEnd < newIntervalStart) {
                outputIntervals.add(interval);
            } else if(currentIntervalStart > newIntervalEnd) {
                
                if(!inserted) {
                    outputIntervals.add(new int[]{newIntervalStart, newIntervalEnd});
                    inserted = true;
                }

                outputIntervals.add(interval);
            } else {
                newIntervalStart = Math.min(newIntervalStart, currentIntervalStart);
                newIntervalEnd = Math.max(newIntervalEnd, currentIntervalEnd);
            }

        }

        if(!inserted) {
            outputIntervals.add(new int[]{newIntervalStart, newIntervalEnd});
        }

        return outputIntervals.toArray(new int[outputIntervals.size()][]);
    }
}

File: interval/insert-interval/insert_interval.py
# https://leetcode.com/problems/insert-interval/?envType=study-plan-v2&envId=top-interview-150
# Time Complexity: O(n) where n is the number of intervals
# Space Complexity: O(n) where n is the number of intervals

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        output = []
        i = 0
        n = len(intervals)
        
        # Phase 1: Add all intervals that come strictly BEFORE newInterval
        while i < n and intervals[i][1] < newInterval[0]:
            output.append(intervals[i])
            i += 1
            
        # Phase 2: Merge all overlapping intervals with newInterval
        while i < n and intervals[i][0] <= newInterval[1]:
            newInterval[0] = min(newInterval[0], intervals[i][0])
            newInterval[1] = max(newInterval[1], intervals[i][1])
            i += 1
        
        # Insert the safely merged newInterval
        output.append(newInterval)
        
        # Phase 3: Add all intervals that come strictly AFTER newInterval
        output.extend(intervals[i:])
            
        return output